exporters/otlp/otlptrace/exporter_test.go

```diff
@@ -0,0 +1,173 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package otlptrace_test
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"testing"
+	"time"
+
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+
+	"github.com/stretchr/testify/assert"
+
+	"go.opentelemetry.io/otel"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	tracepb "go.opentelemetry.io/proto/otlp/trace/v1"
+)
+
+type stubClient struct {
+	rs []*tracepb.ResourceSpans
+}
+
+var _ otlptrace.Client = (*stubClient)(nil)
+
+func (m *stubClient) Start(_ context.Context) error {
+	return nil
+}
+
+func (m *stubClient) Stop(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+	return nil
+}
+
+func (m *stubClient) UploadTraces(ctx context.Context, protoSpans []*tracepb.ResourceSpans) error {
+	m.rs = append(m.rs, protoSpans...)
+	return nil
+}
+
+func (m *stubClient) Reset() {
+	m.rs = nil
+}
+
+func TestExporterShutdownHonorsTimeout(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	e := otlptrace.NewUnstartedExporter(&stubClient{})
+	if err := e.Start(ctx); err != nil {
+		t.Fatalf("failed to start exporter: %v", err)
+	}
+
+	innerCtx, innerCancel := context.WithTimeout(ctx, time.Microsecond)
+	<-time.After(time.Second)
+	if err := e.Shutdown(innerCtx); err == nil {
+		t.Error("expected context DeadlineExceeded error, got nil")
+	} else if !errors.Is(err, context.DeadlineExceeded) {
+		t.Errorf("expected context DeadlineExceeded error, got %v", err)
+	}
+	innerCancel()
+}
```

Comment:
Does this test that the Exporter honors shutdown, or that the client does?
For example, if `stubClient` just ignored the `ctx` and spun will it still shutdown?

I know this is a copy, but we might want to move it to the client, or remove it.

Reply:
Yes, you are correct. In the past this tested the connection direct, and at the point that the `ProtocolDriver` interface was added, this started to test a stub client, which doesn't really test the actual code. 

I've moved that to the client end-to-end tests, and will be test on both gRPC/HTTP clients.